sda-commons-server-dropwizard/src/main/java/org/sdase/commons/server/dropwizard/metadata/MetadataContext.java

```diff
@@ -0,0 +1,119 @@
+package org.sdase.commons.server.dropwizard.metadata;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ * The {@code MetadataContext} stores information about a long-running business processes that is
+ * independent of specific APIs or business rules of a single service. A {@code MetaContext} is
+ * built from the metadata fields of input APIs like HTTP of Kafka message headers. It is added
+ * automatically to output APIs when using platform clients from the {@code JerseyClientBundle} or
+ * the {@code KafkaMessageProducer}.
+ *
+ * <p>Services that claim to support {@code MetadataContext} must take care that it is also kept and
+ * reloaded when the business process is interrupted and proceeded later. An interrupted business
+ * process in these terms is for example asynchronous processing in a new {@link Thread} or
+ * finishing the current step of the process by saving the current state to a database.
+ *
+ * <p>The metadata fields (e.g. header names) that are put into the {@code MetadataContext} must be
+ * configured by {@linkplain System#getenv(String) environment variable} or {@linkplain
+ * System#getProperty(String) system property} {@code METADATA_HEADERS} as comma separated list of
+ * header names, e.g. {@code business-process-id,tenant-id}. Header names are treated
+ * case-insensitive.
+ *
+ * <p>Technically, the {@code MetadataContext} is implemented as static {@link ThreadLocal}. When
+ * code is executed in a new {@link Thread}, the context must be transferred. Helper methods are
+ * available for {@linkplain #transferMetadataContext(Runnable) <code>Runnable</code>} and
+ * {@linkplain #transferMetadataContext(Callable) <code>Callable</code>}
+ */
+// TODO not picked up from request headers and Kafka consumers as documented yet.
+// TODO Not handled in platform client and KafkaMessageProducer as documented yet.
+public interface MetadataContext {
+
+  /**
+   * @return the current immutable metadata context, never {@code null}.
+   */
+  static MetadataContext current() {
+    return MetadataContextUtil.current();
+  }
+
+  /**
+   * @return the current metadata context, never {@code null}. Changes in the returned instance will
+   *     not affect the {@link MetadataContext} of the current {@link Thread}.
+   */
+  static DetachedMetadataContext detachedCurrent() {
+    return DetachedMetadataContext.of(MetadataContextUtil.current());
+  }
```

Comment:
We could add `@NotNull` to the return type. Would make it easier for IDEs to find possible bugs. We've never discussed if we want to make use of that approach AFAIK.

Reply:
I'm not a big fan of annotating the code that way.

Reply:
Then let's not start with it. 😄